source/validate.h

```diff
@@ -36,33 +46,267 @@
+  std::vector<std::vector<uint32_t>> parameter_ids_;
+
+  // NOTE: See correspoding getter functions
+  bool in_function_;
+  bool in_block_;
+};
+
+class ValidationState_t {
+ public:
+  ValidationState_t(spv_diagnostic* diagnostic, uint32_t options);
+
+  // Forward declares the id in the module
+  spv_result_t forwardDeclareId(uint32_t id);
+
+  // Removes a forward declared ID if it has been defined
+  spv_result_t removeIfForwardDeclared(uint32_t id);
+
+  // Assigns a name to an ID
+  void assignNameToId(uint32_t id, std::string name);
+
+  // Returns a string representation of the ID in the format <id>[Name] where
+  // the <id> is the numeric valid of the id and the Name is a name assigned by
+  // the OpName instruction
+  std::string getIdName(uint32_t id) const;
+
+  // Returns the number of ID which have been forward referenced but not defined
+  size_t unresolvedForwardIdCount() const;
+
+  // Returns a list of unresolved forward ids.
+  std::vector<uint32_t> unresolvedForwardIds() const;
+
+  // Returns true if the id has been defined
+  bool isDefinedId(uint32_t id) const;
+
+  // Returns true if an spv_validate_options_t option is enabled in the
+  // validation instruction
+  bool is_enabled(spv_validate_options_t flag) const;
+
+  // Increments the instruction count. Used for diagnostic
+  int incrementInstructionCount();
+
+  // Returns the current layout section which is being processed
+  ModuleLayoutSection getLayoutSection() const;
+
+  // Increments the module_layout_order_section_
+  void progressToNextLayoutSectionOrder();
+
+  // Determines if the op instruction is part of the current section
+  bool isOpcodeInCurrentLayoutSection(SpvOp op);
+
+  libspirv::DiagnosticStream diag(spv_result_t error_code) const;
+
+  // Returns the function states
+  Functions& get_functions();
+
+  // Retuns true if the called after a function instruction but before the
+  // function end instruction
+  bool in_function_body() const;
+
+  // Returns true if called after a label instruction but before a branch
+  // instruction
+  bool in_block() const;
+
+  // Keeps track of ID definitions and uses.
+  class UseDefTracker {
+   public:
+    void AddDef(const spv_id_info_t& def) { defs_.push_back(def); }
+
+    void AddUse(uint32_t id) { uses_.insert(id); }
+
+    // Finds id's def, if it exists.  If found, returns <true, def>.  Otherwise,
+    // returns <false, something>.
+    std::pair<bool, spv_id_info_t> FindDef(uint32_t id) const {
+      auto found =
+          std::find_if(defs_.cbegin(), defs_.cend(),
+                       [id](const spv_id_info_t& e) { return e.id == id; });
+      if (found == defs_.cend()) {
+        return std::make_pair(false, spv_id_info_t{});
+      } else {
+        return std::make_pair(true, *found);
+      }
+    }
+
+    // Returns uses of IDs lacking defs.
+    std::unordered_set<uint32_t> FindUsesWithoutDefs() const {
+      auto diff = uses_;
+      for (const auto d : defs_) diff.erase(d.id);
```

Comment:
It looks like the ValidateID.cpp tests require the SPV_VALIDATE_SSA_BIT flag to work correctly. These flags are getting unnecessary complicated. Perhaps we need to run the tests by directly calling the function instead of spvValidate + flag.

I should also further document the responsibilities of each of the passes.